File: examples/tempopod.js
// Use: node tempopod.js 45 https://raw.githubusercontent.com/webreactiva-devs/reto-tempopod/main/feed/webreactiva.xml
// Use: node tempopod.js 

const args = process.argv.slice(2);
const selectedTempo = args[0] || 30; // Minutes
const feedUrl =
  args[1] ||
  "https://raw.githubusercontent.com/webreactiva-devs/reto-tempopod/main/feed/webreactiva.xml";

fetch(feedUrl)
  .then((response) => response.text())
  .then((str) => {
    const episodes = [];
    const itemRegex = /<item>(.*?)<\/item>/gs;
    let itemMatch;
    while ((itemMatch = itemRegex.exec(str)) !== null) {
      const episodeContent = itemMatch[1];
      const titleMatch = episodeContent.match(/<title>(.*?)<\/title>/);
      const durationMatch = episodeContent.match(
        /<itunes:duration>(.*?)<\/itunes:duration>/
      );
      if (titleMatch && durationMatch) {
        episodes.push({
          title: titleMatch[1],
          duration: parseInt(durationMatch[1]),
        });
      }
    }
    return selectEpisodes(episodes, selectedTempo);
  })
  .then((selectedEpisodes) =>
    console.log("Episodios seleccionados:", selectedEpisodes)
  )
  .catch((error) => console.error("Error:", error));

function selectEpisodes(episodes, tempo) {
  let selected = [];
  let totalTime = 0;
  episodes = episodes.sort(() => 0.5 - Math.random());

  for (let episode of episodes) {
    if (totalTime + episode.duration <= tempo * 60) {
      selected.push(episode.title);
      totalTime += episode.duration;
    }
  }

  return selected;
}

File: examples/TempoPod.java
// Tested :)
// ~/test via ☕ v17.0.10
// ➜ java TempoPod 180

import javax.xml.parsers.*;
import org.w3c.dom.*;
import org.xml.sax.InputSource;
import java.io.StringReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TempoPod {

    public static void main(String[] args) throws Exception {
        int selectedTempo = args.length > 0 ? Integer.parseInt(args[0]) : 30;
        String feedUrl = args.length > 1 ? args[1] : "https://raw.githubusercontent.com/webreactiva-devs/reto-tempopod/main/feed/webreactiva.xml";

        String xmlContent = fetchFeed(feedUrl);
        List<Episode> episodes = parseEpisodes(xmlContent);
        List<String> selectedEpisodes = selectEpisodes(episodes, selectedTempo);
        System.out.println("Episodios seleccionados:");
        selectedEpisodes.forEach(System.out::println);
    }

    private static String fetchFeed(String feedUrl) throws Exception {
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(feedUrl))
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        return response.body();
    }

    private static List<Episode> parseEpisodes(String xmlContent) throws Exception {
        List<Episode> episodes = new ArrayList<>();
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        Document document = builder.parse(new InputSource(new StringReader(xmlContent)));
        NodeList items = document.getElementsByTagName("item");
        for (int i = 0; i < items.getLength(); i++) {
            Node item = items.item(i);
            if (item.getNodeType() == Node.ELEMENT_NODE) {
                Element element = (Element) item;
                String title = element.getElementsByTagName("title").item(0).getTextContent();
                String durationStr = element.getElementsByTagName("itunes:duration").item(0).getTextContent();
                int duration = Integer.parseInt(durationStr);
                episodes.add(new Episode(title, duration));
            }
        }
        return episodes;
    }

    private static List<String> selectEpisodes(List<Episode> episodes, int tempo) {
        Collections.shuffle(episodes);
        List<String> selected = new ArrayList<>();
        int totalTime = 0;
        for (Episode episode : episodes) {
            if (totalTime + episode.duration <= tempo * 60) {
                selected.add(episode.title);
                totalTime += episode.duration;
            }
        }
        return selected;
    }

    static class Episode {
        String title;
        int duration;

        Episode(String title, int duration) {
            this.title = title;
            this.duration = duration;
        }
    }
}
